Replace pkg_resources

rokm opened the pull request:
Replace the use of deprecated `pkg_resources` throughout the codebase with combination of `importlib.metadata` / `importlib_metadata` and `packaging`.

Closes #6458.

rokm committed:
building: use importlib.metadata for discovery of hook directories

Switch to `importlib.metadata` (or `importlib_metadata`) for entry
point enumeration during the discovery of hook directories.

rokm committed:
tests: add tests for run-time use of metadata collected from .egg

Add two test that try to query metadata that was collected from
(non-zipped) .egg. In such cases, we collect the EGG-INFO directory
containing the metadata under its original parent directory
(the package_name-version.egg) to prevent clashes between
metadata from multiple .egg sources. The first test queries the
metadata using `pkg_resources`, while the second uses
`importlib.metadata` / `importlib_metadata`.

rokm committed:
bootstrap: add .egg directories in sys._MEIPASS to sys.path

It seems that `pkg_resources` is able to discover metadata
collected from eggs as long as the parent directory of the
.egg directory is in `sys.path`. This is automatically true in
our case, because we collect the  `package_name-version.egg`
directories (that contain `EGG-INFO` metadata directories) into
`sys._MEIPASS`.

In contrast, `importlib.metadata` does not seem to pick up
such metadata unless the `.egg` directories themselves are
in `sys.path`. So have the bootstrap script scan for directories
with `.egg` suffix and append them so `sys.path`.

rokm committed:
hookutils: port metadata collection to importlib.metadata

Port metadata collection (`copy_metadata` hook utility function)
from `pkg_resources` to `importlib.metadata` and `packaging`.
The latter is required for parsing package requirements and
evaluating markers to exclude the requirements that are not
applicable to the running python version and/or platform.

Simplify the metadata destination handling code; the only
special case we need to care about is EGG-INFO directory in
.egg  directories, where we need to preserve the parent
directory instead of collecting into top-level application
directory.

`importlib.metadata` seems to properly resolve the path to
egg-info files used by Debian/Ubuntu packages, so we can also
remove the legacy name/path handling.

Have the remaining hook utility functions that make use of
`pkg_resources` import it locally, so we can port them one-by-one.

rokm committed:
hookutils: port get_installer to importlib.metadata

Port the `get_installer` helper from `pkg_resources` to
`importlib.metadata`. Requires python >= 3.10 or equivalent
`importlib-metadata` (>= 4.6) due to use of the
`packages_distributions()` convenience function.

rokm committed:
analysis: use importlib.metadata in error handling in _metadata_from

Since the `copy_metadata` calls in `_metadata_from` now use
`importlib.metadata` instead of `pkg_resources`, we now need
to handle the `importlib.metadata.PackageNotFoundError` instead
of `pkg_resources.PackageNotFoundError`.

Use `compat.importlib_metadata` to properly resolve
`importlib.metadata` vs `importlib_metadata` regardless if the
latter is available on python > 3.10, where our `compat` prefers
the stdlib version.

rokm committed:
hookutils: port collect_entry_point to importlib.metadata

rokm committed:
compat: resolve importlib.metadata vs importlib_metadata

Import `importlib.metadata` as `importlib_metadata` if stdlib
version is recent enough (python >= 3.10). Otherwise import
the `importlib_metadata`, but explicitly verify the version
of `importlib-metadata` dist, which must be >= 4.6. This
version is needed for extended functionality, such as `group`
and `name` keyword arguments having been added to the
`entry_points()` function.

Add `importlib-metadata >= 4.6` to requirements if python < 3.10.

Add `packaging` >= 20.0 to requirements; used for parsing of
version and requirements strings.

rokm committed:
compat: generalize setup.py hack for missing run-time dependencies

Replace the `PYINSTALLER_NO_PYWIN32_FAILURE` environment variable
set by `setup.py` into more general `_PYINSTALLER_SETUP_PY`, which
now covers the following run-time dependencies in the `compat`
module:
* `importlib_metadata` on python < 3.10
* `pywin32-ctypes` on Windows

This "setup.py mode" prevents `compat` from trying to import
afore-mentioned run-time dependencies, which may not be present
when PyInstaller wheels are being built (nor are required for
the `compat` functionality used by the `setup.py`).

rokm committed:
hookutils: port requirements_for_package() to importlib.metadata

rokm committed:
hookutils: have collect_all look up the dist from package name

Have `collect_all` look up the dist name for the given package
name before attempting to call `copy_metadata` and
`requirements_for_package`. This fixes the collection (and
spurious warnings) for cases when dist name does not match the
importable/package name.

rokm committed:
hookutils: adjust behavior of collect_data_files with include_py_files

Adjust the behavior of `collect_data_files` when the `include_py_files`
flag is enabled: collect only `.py` and `.pyc` files, and never
collect `.pyc` files from the `__pycache__` directory.

Update the description to note that module collection mode is
preferable way of ensuring that source .py files are collected
(in addition or in lieu to byte-compiled modules in PYZ).

Adjust the `test_collect_data_all_included` accordingly.

rokm committed:
modulegraph: remove the modulegraph.replacePackage function

The only use for `modulegraph.replacePackage` is for the `_xmplus`
module from python2-era PyXML package. Remove the function and
associated code.

rokm committed:
modulegraph: rework support for legacy setuptools namespace packages

Rework the support for legacy setuptools-based namespace packages
and port it to `importlib.metadata`.

The legacy namespace packages are regular packages with `__init__.py`
that calls deprecated  `pkg_resources.declare_namespace(__name__)`.

We could get rid of this altogether once we are fine with not
supporting legacy namespace packages that are split across
multiple locations. Currently, our test suite contains such an
example (e.g., `test_import.py::test_nspkg1`). The code has been
refactored so that dropping support will also be easier.

To properly support legacy namespace sub-packages that are split
across different locations, we need to scan the metadata of all
dists, similarly to how it was done by the original `pkg_resources`
based implementation. However, within the life-time of a module
graph instance, we can perform the scan only once, and cache the
results.

rokm committed:
modulegraph: remove parsing of setuptools' *-nspkg.pth files

In contemporary python, namespace packages should be handled without
having to resort to parsing the *-nspkg.pth files.

rokm committed:
utils: run_tests: use importlib.metadata to discover tests

Port the entry-point scan for test discovery from `pkg_resources`
to `importlib.metadata`.

rokm committed:
hookutils: implement replacement for is_module_satisfies

Replace `is_module_satisfies` with `check_requirement`, which
uses `packaging` to parse the requirement and check the version
against the optional specifier(s), and `importlib.metadata` to
query the dist version.

Keep `is_module_satisfies` as an alias for `check_requirements`
to maintain compatibility with existing hooks (from the contrib
hooks repository) that call the function with only the
`requirement` string. The calls with additional `version` and
`version_attr` arguments will not work anymore, nor will the
fall back to checking module's version when the specified
requirement is not satisfied.

Have the @requires test decorator use `check_requirement`.

Update hooks in this repository to use `check_requirement`
instead of `is_module_satisfies`.

Update Qt hook utilities, where we need to account for the
fact that `check_requirement` does not attempt to fall back
to checking module's version (which in this case was
undesirable and had to be worked around).

bwoodsend commented on PyInstaller/compat.py:
```diff
@@ -24,6 +24,23 @@
+    import packaging.version  # For importlib_metadata version check
+
+    # Validate the version
+    if packaging.version.parse(importlib_metadata.version("importlib-metadata")) < packaging.version.parse("4.6"):
```
The version in the `setup.cfg` should be enough to ensure that this never happens. At least I think it should.

rokm replied:
True, but that is only at PyInstaller's installation time. Nothing really prevents a determined user (or possibly a misconfigured package) from installing a different version of `importlib-metadata` later on. And I would prefer to shut down such nonsense as soon as possible instead of dealing with tracebacks about `importlib_metadata` missing `packages_distributions` or `importlib_metadata.entry_points` not having a `groups` argument.

bwoodsend commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1016,44 +1016,40 @@ def get_installer(module: str):
+            installer_text = dist.read_text('INSTALLER')
+            if installer_text is not None:
+                return installer_text.strip()
+        except importlib_metadata.PackageNotFoundError:
```
Can we add `FileNotFoundError` (assuming that that is the exception type if the `INSTALLER` file is missing) to the permitted exceptions. I don't trust every Linux package manager to have never tried to stitch this metadata together themselves and missed bits.

rokm replied:
If the file is missing, the `read_text` method should return `None`.

```
Help on function read_text in module importlib_metadata:

read_text(self, filename)
    Attempt to load metadata file given by the name.
    
    :param filename: The name of the file in the distribution info.
    :return: The text if found, otherwise None.
```

bwoodsend commented on PyInstaller/utils/hooks/__init__.py:
```diff
@@ -1053,65 +1052,35 @@ def get_installer(module: str):
+    def _top_level_inferred(dist):
+        return [
+            f.parts[0] if len(f.parts) > 1 else f.with_suffix('').name for f in (dist.files or []) if f.suffix == ".py"
+        ]
```
Could this mapping of distributions to modules not also be done by looking at `importlib.metadata.packages_distributions()`?

rokm replied:
It could, although `importlib.metadata.packages_distributions()` is exactly the opposite of what we need here (we'd need to invert the dictionary).

I think the only known caller of this function is `collect_all`, which needs `importlib.metadata.packages_distributions()` dictionary anyway; so it might make more sense to also invert it and perform the lookup there.

Or perhaps even drop this module -> dist -> requirements -> top-level-modules lookup  altogether.

bwoodsend replied:
I've been meaning to scrap the requirements component of `collect_all()` for ages.

rokm replied:
Let's get rid of it here, then.

bwoodsend commented on news/7943.breaking.rst:
I think we need a feature entry too to announce the existence of/version added for `check_requirement()`

rokm replied:
Yeah, that makes sense.

rokm committed:
hookutils: remove requirements_for_package

Remove `requirements_for_package` hook utility, which was
used by `collect_all`. Consequently, the latter does not include
the top-level modules of metadata-declared requirements among the
returned hidden imports anymore.

bwoodsend commented:
Guess we can close #6458 then too.
